OAuth 2 Tests depend on dict ordering

URL strings may be generated differently depending on the internal dict order which is fine. Our tests however do naive string comparisons which is not adequate and breaks in python 3.3. There might already be a function such as assertURLEquals but I've not looked into this. If not it is worth monkey patching this onto TestCase and using it.  It should be quite straightforward using urlsparse and just making sure all query and fragment pairs are present in both strings.

Test output

```
======================================================================
FAIL: test_request_body (tests.oauth2.draft25.test_client.WebApplicationClientTest)
----------------------------------------------------------------------
Traceback (most recent call last):
    File "/home/ib/Testing/oauthlib/tests/oauth2/draft25/test_client.py", line 181, in test_request_body
      self.assertEqual(body, self.body_redirect)
AssertionError: 'not=empty&grant_type=authorization_code&redirect_uri=http%3A%2F%2Fmy.page.com%2 [truncated]... != 'not=empty&grant_type=authorization_code&code=zzzzaaaa&redirect_uri=http%3A%2F%2 [truncated]...
- not=empty&grant_type=authorization_code&redirect_uri=http%3A%2F%2Fmy.page.com%2Fcallback&code=zzzzaaaa
?                                                                                         --------------
+ not=empty&grant_type=authorization_code&code=zzzzaaaa&redirect_uri=http%3A%2F%2Fmy.page.com%2Fcallback
?                                        ++++++++++++++


======================================================================
FAIL: Verify correct access token request body construction.
----------------------------------------------------------------------
Traceback (most recent call last):
    File "/home/ib/Testing/oauthlib/tests/oauth2/draft25/test_parameters.py", line 124, in test_prepare_token_request
      self.assertEqual(prepare_token_request(**self.grant_body), self.auth_grant_body)
AssertionError: 'grant_type=authorization_code&redirect_uri=https%3A%2F%2Fclient.example.com%2Fc [truncated]... != 'grant_type=authorization_code&code=SplxlOBeZQQYbYS6WxSbIA&redirect_uri=https%3A [truncated]...
- grant_type=authorization_code&redirect_uri=https%3A%2F%2Fclient.example.com%2Fcb&code=SplxlOBeZQQYbYS6WxSbIA
+ grant_type=authorization_code&code=SplxlOBeZQQYbYS6WxSbIA&redirect_uri=https%3A%2F%2Fclient.example.com%2Fcb


======================================================================
FAIL: test_dict_body (tests.test_common.CommonTests)
----------------------------------------------------------------------
Traceback (most recent call last):
    File "/home/ib/Testing/oauthlib/tests/test_common.py", line 82, in test_dict_body
      self.assertEqual(r.decoded_body, self.params_twotuple)
AssertionError: Lists differ: [('baz', '123'), ('foo', 'bar'... != [('foo', 'bar'), ('baz', '123'...

First differing element 0:
('baz', '123')
('foo', 'bar')

- [('baz', '123'), ('foo', 'bar')]
+ [('foo', 'bar'), ('baz', '123')]

======================================================================
FAIL: test_extract_params_dict (tests.test_common.CommonTests)
----------------------------------------------------------------------
Traceback (most recent call last):
    File "/home/ib/Testing/oauthlib/tests/test_common.py", line 30, in test_extract_params_dict
      self.assertEqual(extract_params(self.params_dict), self.params_twotuple)
AssertionError: Lists differ: [('baz', '123'), ('foo', 'bar'... != [('foo', 'bar'), ('baz', '123'...

First differing element 0:
('baz', '123')
('foo', 'bar')

- [('baz', '123'), ('foo', 'bar')]
+ [('foo', 'bar'), ('baz', '123')]

----------------------------------------------------------------------
```

--- a/tests/oauth2/draft25/test_client.py
+++ b/tests/oauth2/draft25/test_client.py
@@ -223,10 +223,9 @@ class WebApplicationClientTest(TestCase):
         uri = client.prepare_request_uri(self.uri, state=self.state)
         self.assertURLEqual(uri, self.uri_state)
 
-        # With extra parameters through kwargs, checking using len since order
-        # of dict items is undefined
+        # With extra parameters through kwargs
         uri = client.prepare_request_uri(self.uri, **self.kwargs)
-        self.assertEqual(len(uri), len(self.uri_kwargs))
+        self.assertURLEqual(uri, self.uri_kwargs)
 
     def test_request_body(self):
         client = WebApplicationClient(self.client_id, code=self.code)
@@ -243,10 +242,9 @@ class WebApplicationClientTest(TestCase):
         body = client.prepare_request_body(body=self.body, redirect_uri=self.redirect_uri)
         self.assertFormBodyEqual(body, self.body_redirect)
 
-        # With extra parameters, checked using length since order of
-        # dict items is undefined
+        # With extra parameters
         body = client.prepare_request_body(body=self.body, **self.kwargs)
-        self.assertEqual(len(body), len(self.body_kwargs))
+        self.assertFormBodyEqual(body, self.body_kwargs)
 
     def test_parse_grant_uri_response(self):
         client = WebApplicationClient(self.client_id)
@@ -328,10 +326,9 @@ class UserAgentClientTest(TestCase):
         uri = client.prepare_request_uri(self.uri, state=self.state)
         self.assertURLEqual(uri, self.uri_state)
 
-        # With extra parameters through kwargs, checking using len since order
-        # of dict items is undefined
+        # With extra parameters through kwargs
         uri = client.prepare_request_uri(self.uri, **self.kwargs)
-        self.assertEqual(len(uri), len(self.uri_kwargs))
+        self.assertURLEqual(uri, self.uri_kwargs)
 
     def test_parse_token_response(self):
         client = UserAgentClient(self.client_id)
@@ -386,11 +383,10 @@ class PasswordCredentialsClientTest(TestCase):
                 body=self.body)
         self.assertFormBodyEqual(body, self.body_up)
 
-        # With extra parameters, checked using length since order of
-        # dict items is undefined
+        # With extra parameters
         body = client.prepare_request_body(self.username, self.password,
                 body=self.body, **self.kwargs)
-        self.assertEqual(len(body), len(self.body_kwargs))
+        self.assertFormBodyEqual(body, self.body_kwargs)
 
     def test_parse_token_response(self):
         client = PasswordCredentialsClient(self.client_id)
@@ -444,10 +440,9 @@ class ClientCredentialsClientTest(TestCase):
         body = rclient.prepare_request_body(body=self.body)
         self.assertFormBodyEqual(body, self.body_up)
 
-        # With extra parameters, checked using length since order of
-        # dict items is undefined
+        # With extra parameters
         body = client.prepare_request_body(body=self.body, **self.kwargs)
-        self.assertEqual(len(body), len(self.body_kwargs))
+        self.assertFormBodyEqual(body, self.body_kwargs)
 
     def test_parse_token_response(self):
         client = ClientCredentialsClient(self.client_id)

